liftedinit/many-rs
# Bazel support

First draft. MVP.

Supported
- Building `many-rs`
- Running `many-rs` tests and doc tests
- Running `many-rs` code coverage
  - ~Currently broken until https://github.com/bazelbuild/rules_rust/pull/1501 is merged and our code is updated to a new release~
- Running `rustfmt` and `clippy`
- Generating the project file required by `rust-analyzer` → https://bazelbuild.github.io/rules_rust/rust_analyzer.html#vscode
  - **Untested** 

TODO
- Activate crate features from the CLI → https://github.com/bazelbuild/rules_rust/issues/801
- Create a release/distribution package → https://github.com/bazelbuild/rules_rust/issues/458
- IntelliJ integration
  - https://github.com/bazelbuild/intellij/issues/675
  - https://github.com/bazelbuild/intellij/pull/3222
  - https://github.com/intellij-rust/intellij-rust/pull/8491
  - JetBrains is co-maintaining the Bazel IntelliJ plugin, and a JetBrains engineer told me that "they should get around to it soon."
- `cargo update`, maybe? → https://github.com/bazelbuild/bazel/issues/10332
- Adjust `clippy` configuration, `rustfmt` configuration, more CLI config, polishing...
- More DRY
- README

To (re-)sync crates: `CARGO_BAZEL_REPIN=1 bazel sync --only=crate_index`
To build: `bazel build //...:all-targets`
To test: `bazel test //...:all-targets`
To run code coverage: 
```
  $ bazel coverage --combined_report=lcov //...:all-targets 
  $ genhtml --output genhtml "$(bazel info output_path)/_coverage/_coverage_report.dat"
```
To run rustfmt: `bazel build --config=rustfmt //...:all-targets`
To run clippy: `bazel build --config=clippy //...:all-targets`
To generate rust-analyzer files: `bazel build --config=rust-analyzer //...:all-targets`
